sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala

```diff
@@ -74,6 +78,43 @@ case class InterpretedMutableProjection(expressions: Seq[Expression]) extends Mu
 }
 
 /**
+ * A projection that could turn any InternalRow into UnsafeRow
+ */
+case class ToUnsafeProjection(fields: Seq[DataType]) extends Projection {
+
+  private[this] val converter = new UnsafeRowConverter(fields.toArray)
+  private[this] val pool = new ObjectPool(10)
+  private[this] var buffer = new Array[Byte](1024)
+  private[this] val unsafeRow = new UnsafeRow()
+
+  override def apply(input: InternalRow): InternalRow = {
+    val numBytes = converter.getSizeRequirement(input)
+    if (numBytes > buffer.length) {
+      buffer = new Array[Byte](numBytes)
+    }
+    pool.clear()
+    converter.writeRow(input, buffer, PlatformDependent.BYTE_ARRAY_OFFSET, numBytes, pool)
```

Comment:
UnsafeColumnWriters are also slow, I think we could have a generated UnsafeRowConverter, sounds good?